iothub/device/devdoc/API Design.md

```diff
@@ -104,27 +108,22 @@ public Task SubscribeToWritablePropertiesEventAsync(Func<ClientPropertyCollectio
 #### All related types
 
 ```csharp
-public class ClientProperties : IEnumerable, IEnumerable<object> {
-    public object this[string key] { get; }
-    public long Version { get; }
+public class ClientProperties : ClientPropertyCollection {
     public ClientPropertyCollection Writable { get; private set; }
-    public bool Contains(string propertyName);
-    public IEnumerator<object> GetEnumerator();
-    IEnumerator System.Collections.IEnumerable.GetEnumerator();
-    public bool TryGetValue<T>(string componentName, string propertyName, out T propertyValue);
-    public bool TryGetValue<T>(string propertyKey, out T propertyValue);
 }
 
 public class ClientPropertyCollection : PayloadCollection {
     public ClientPropertyCollection();
-    public long Version { get; private set; }
+    public long Version { get; protected set; }
     public void Add(IDictionary<string, object> properties, string componentName = null);
     public override void Add(string propertyName, object propertyValue);
+    public void Add(string propertyName, object propertyValue, int statusCode, long version, string description = null, string componentName = null);
     public void Add(string propertyName, object propertyValue, string componentName);
-    public void AddOrUpdate(IDictionary<string, object> properties, string componentName = null);
+    public void AddOrUpdate(IDictionary<string, object> properties, string componentNam = null);
     public override void AddOrUpdate(string propertyName, object propertyValue);
+    public void AddOrUpdate(string propertyName, object propertyValue, int statusCode, long version, string description = null, string componentName = null);
```

Comment:
**NOTE**

Did we want to consider putting the component name first for all of our Add methods? This way the flow is hierarchical.